contracts/pcv/utils/ERC20Allocator.sol

```diff
@@ -0,0 +1,450 @@
+/// This contract should never hold PCV, however it has a sweep function, so if tokens get sent to it accidentally,
+/// they can still be recovered.
+
+/// This contract stores each PSM and maps it to the target balance and decimals normalizer for that token
+/// PCV Deposits can then be linked to these PSM's which allows funds to be pushed and pulled
+/// between these PCV deposits and their respective PSM's.
+/// This design allows multiple PCV Deposits to be linked to a single PSM.
+
+/// This contract enforces the assumption that all pcv deposits connected to a given PSM share
+/// the same underlying token, otherwise the rate limited logic will not work as intended.
+/// This assumption is encoded in the create and edit deposit functions as well as the
+/// connect deposit function.
+
+/// @author Elliot Friedman
+contract ERC20Allocator is IERC20Allocator, CoreRef, RateLimitedV2 {
+    using Address for address payable;
+    using SafeERC20 for IERC20;
+    using SafeCast for *;
+
+    /// @notice container that stores information on all psm's and their respective deposits
+    struct PSMInfo {
+        /// @notice target token address to send
+        address token;
+        /// @notice only skim if balance of target is greater than targetBalance
+        /// only drip if balance of target is less than targetBalance
+        uint248 targetBalance;
+        /// @notice decimal normalizer to ensure buffer is updated uniformly across all deposits
+        int8 decimalsNormalizer;
+    }
+
+    /// @notice map the psm address to the corresponding target balance information
+    /// excess tokens past target balance will be pulled from the PSM
+    /// if PSM has less than the target balance, tokens will be sent to the PSM
+    mapping(address => PSMInfo) public allPSMs;
+
+    /// @notice map the pcv deposit address to a peg stability module
+    mapping(address => address) public pcvDepositToPSM;
+
+    /// @notice ERC20 Allocator constructor
+    /// @param _core Volt Core for reference
+    /// @param _maxRateLimitPerSecond maximum rate limit per second that governance can set
+    /// @param _rateLimitPerSecond starting rate limit per second
+    /// @param _bufferCap cap on buffer size for this rate limited instance
+    constructor(
+        address _core,
+        uint256 _maxRateLimitPerSecond,
+        uint128 _rateLimitPerSecond,
+        uint128 _bufferCap
+    )
+        CoreRef(_core)
+        RateLimitedV2(_maxRateLimitPerSecond, _rateLimitPerSecond, _bufferCap)
+    {}
+
+    /// ----------- Governor Only API -----------
+
+    /// @notice connect a new PSM
+    /// @param psm Peg Stability Module to add
+    /// @param targetBalance target amount of tokens for the PSM to hold
+    /// @param decimalsNormalizer decimal normalizer to ensure buffer is depleted and replenished properly
+    function connectPSM(
+        address psm,
+        uint248 targetBalance,
+        int8 decimalsNormalizer
```

Comment:
do we need to pass in the decimalsNormalizer? would it possible to just have parity with the decimalsNormalizer for the PSM

Reply:
we could, but that would require changing code even more which I am reluctant to do at this point.